File: system/BasedCopable.java
package common.system;

import common.io.assets.Admin.StaticPermitted;

import java.util.HashMap;
import java.util.Map;

public interface BasedCopable<T, B> extends Cloneable, Copable<T> {

    @StaticPermitted(StaticPermitted.Type.TEMP)
    Map<Class<?>, Object> map = new HashMap<>();

    @Override
    @SuppressWarnings("unchecked")
    default T copy() {
        B base = (B) map.get(getClass());
        if (base == null)
            return null;
        return copy(base);
    }

    T copy(B b);

}

File: system/BasedCopable.kt
package common.system

import common.io.assets.Admin
import common.io.assets.Admin.StaticPermitted
import java.util.*

interface BasedCopable<T, B> : Cloneable, Copable<T?> {
    override fun copy(): T? {
        val base = map[javaClass] as B? ?: return null
        return copy(base)
    }

    fun copy(b: B): T

    companion object {
        @StaticPermitted(Admin.StaticPermitted.Type.TEMP)
        val map: Map<Class<*>, Any> = HashMap()
    }
}
